### PR TITLE
metadata view: introduce a minimal task metadata view

### LINKED ISSUE #1071: task metadata view
**Cylc 7**

Cylc supports task metadata e.g:

```ini
[runtime]
    [[mytask]]
        [[[meta]]]
            # the standard metadata fields used by some cylc commands
            title = My Task
            url = https://...
            description = """
                Lorel ipsum ...
            """
            # you can create custom fields too
            system = xyz
            group = abc
```

Cylc 7 made minimal use of task metadata, namely in commands like `cylc show` but some metadata was also used in the GUI.

**Cylc 8**

Cylc 8 has greater potential here. Earlier design sketches had an info "pane", "tounge" or "panel" which was intended to show the `[meta]` info for the selected task (or the workflow itself as workflows also support `[meta]`). The GUI has moved on since then, the  Lumino tab system is now the natural place for displays, I think dividing the screen beyond this would be unhelpful.

So here's my suggestion, we implement a "metadata view" (or "info" view). This view gives you a text box to type in the name of the task (we should be able to make this a combo box with all task names pre-loaded). Once you've selected a task it displays the `[meta]`. For convenience we can create a shortcut from the task menu which opens a new "info" view pre-initalised at the specified task. We should also consider adding the "title" field to the task menu.

**Sketch**

Here we have an info view showing some basic metadata. The task menu has been opened for the task "baz" on the right. The "title" has been added to the top of the menu along with an info button which is the shortcut for opening the info tab on the left.

task-metadata

**Metadata & Formats**

Some sites are already making use of the "URL" field. In [at least] one case, using HTML pages generated from markdown. We can make this much easier (rendering markdown in a web app is trivial).

* Suggest rendering `[meta]` items as markdown by default.

> This part is more cylc-flow / cylc-rose focused but bringing it up here for context

Many tasks already have extensive documentation in the form of Rose application metadata. There is an existing Rose config to ReST generator which is used by the Rose documentation.

* Consider adding a `[meta]_markup_format` config to allow this to be configurable in the future?
* Or should this be a marker in the content itself e.g. `#markdown`, `#rest`, ...?
* Consider a cylc-rose plugin to drag in Rose application metadata where available.

**Workflows & Documentation**

Workflows can have `[meta]` too. We might want to add an option for viewing the workflow metadata here too (and possibly a shortcut for it in the workflow toolbar).

Cylc has documentation too! We might want to add an option view viewing it here!

**Additional context**

Once discussed / agreed this should be quite quick to implement.

* List of task names can be obtained by listing taskdefs (called `Tasks` in the schema).
* Metadata can be obtained from the taskdef.

**Pull requests welcome!**

### LINKED ISSUE #1896: info view: add completion expression info
Add completion expression information into the "Info View".

See the "completion" section of the `cylc show` command to see what this should include.

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #1071
Closes #1896

* Bare bones task metadata view.
* Display task metadata, prerequisites and outputs.
* Display / functionality improvements and support for other information to follow in future PRs.
 
Screenshot from 2024-08-08 15-44-54

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [ ] Cylc-Doc - bumped changes entry to a [PR](https://github.com/cylc/cylc-doc/issues/756) (view is likely to change further before release)
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by hjoliver** [comment]
Ping @ScottWales @ColemanTom (you were wanting this for viewing prerequisites etc.)

**Comment by ScottWales** [comment]
Great! thanks for this

**Comment by oliver-sanders** [comment]
Added offline-data and tests.

Many, many enhancements and features left to add to this view, but will leave it at basic metadata, prereqs and outputs for this PR.

Further enhancements will slot into this template fairly easily provided the required data is present (and accurate) in the schema.

**Comment by oliver-sanders** [comment]
Documented desired future enhancements in issues:

* https://github.com/cylc/cylc-ui/issues/1893
* https://github.com/cylc/cylc-ui/issues/1894
* https://github.com/cylc/cylc-ui/issues/1895
* https://github.com/cylc/cylc-ui/issues/1896
* https://github.com/cylc/cylc-ui/issues/1897
* https://github.com/cylc/cylc-ui/issues/1898

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:25]: It doesn't on Chrome

image

Comment (oliver-sanders) [reply src/components/cylc/Info.vue:25]: Thanks for testing, will take a look.

Comment (oliver-sanders) [reply src/components/cylc/Info.vue]: Resolved by setting `overflow-x: hidden`.

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:33]: This is not so good for long task names on a narrow window.

Happy to bump to a follow-up issue however

Comment (wxtim) [reply src/components/cylc/Info.vue:33]: Follow up pls.

Comment (wxtim) [reply src/components/cylc/Info.vue:33]: I think this should be a follow up issue.

Comment (oliver-sanders) [reply src/components/cylc/Info.vue:33]: Besides making the icon smaller, I don't think there is actually very much we can do about this. Word wrapping is not an option here.

However, https://github.com/cylc/cylc-ui/issues/1898 will add inputs to allow the user to manually enter the task & cycle similar to the log view. These will likely use smaller font so may display more of the task name, although the log view task input will also get cut off after a point.

File: src/components/cylc/Info.vue
Comment (wxtim) [inline src/components/cylc/Info.vue:214]: I think that this makes more sense: Each output can be produced or not, where as to my mind the tick/cross suggests a more succeed/failed outcome.
```suggestion
      content: '☐';
```

File: src/components/cylc/Info.vue
Comment (wxtim) [inline src/components/cylc/Info.vue:217]: I think that this makes more sense: Each output can be produced or not, where as to my mind the tick/cross suggests a more succeed/failed outcome.
```suggestion
      content: '☐';
```

```suggestion
    .condition:before {
      content: '☐';
      padding-right: 0.5em;
      color: rgb(0, 0, 0);
    }
```

File: src/components/cylc/Info.vue
Comment (wxtim) [inline src/components/cylc/Info.vue:51]: IMO most users will be coming here for the pre-reqs and outputs and not the metadata - these should go first and probably not be collapsed. Could they go in neighbouring Columns?

Happy to spin this out into an issue and not block this PR.

**Review by wxtim — APPROVED** [review]
- [x] Manually tested
- [x] Read PR content

I've made some comments, none of which should block the PR going in if they cost any effort to implement.

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue:217]: I don't think an empty box is clearer than a cross?

This view copies the format used in the scheduler logs and the `cylc view` command.

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue:51]: Columns aren't an option (or at least aren't always an option) because they won't work in narrow widths.

We can't expand all of the sections by default as there will be a lot of them, from prerequisites, to completion expression satisfaction, to runtime configuration and broadcasts. Prereqs are important, but so are all the other sections.

We could start with nothing expanded, but I thought the title, description and URL (often used to link to documentation) are notably important and deserve to be displayed at all times. Important messages like "do not re-run this task" or "if it fails, try re-running in shortstep mode" will be prominent as a result.

File: src/components/cylc/Info.vue
Comment (MetRonnie) [reply src/components/cylc/Info.vue:33]: I don't think we should necessarily use the `GraphNode` component to display the task name and jobs, we can do this more bespoke for this view. But happy to accept the quick win here and punt this to either #1898 or a new issue

Comment (oliver-sanders) [reply src/components/cylc/Info.vue:33]: Aw, I quite like the `GraphNode`, it gives easy access to the task and jobs in a familiar format.

Comment (oliver-sanders) [reply src/components/cylc/Info.vue:33]: Haven't really thought through #1898 yet, we could lift-and-shift the task/workflow selectors from the log view as we need much the same thing here, e.g. this view should display both workflow and task metadata.

File: src/components/cylc/Info.vue
Comment (MetRonnie) [reply src/components/cylc/Info.vue:217]: I agree with Tim's point about green/red being interpreted as success/error.

In the screenshot it looks like the `failed` output is a good thing, while 

image

Also would good to differentiate expected outputs, e.g. by greying out `expired`, `submit-failed` etc. if they are not completed

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:40]: Something is going wrong with the command menu. It does not know what type of Cylc object it is representing (task or cycle or workflow etc.)

image

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:98]: The prereq expression is a little confusing - why is `c` in it? 

image

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue:217]: We don't have required / optional information available via GraphQL at the moment, so all we can do is report if an output has been completed or not. We cannot determine whether the completion or non-completion of an output is a good or a bad thing without that information

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue:40]: Curious, will take a look.

I'm guessing the graph view doesn't show this issue?

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue:98]: This is (currently) how Cylc represents conditional expressions. Please try using the `cylc show` command which this view is replicating.

At the moment, this is the only way the GUI can represent a conditional expression. I have opened an issue to look at improving this, see #1893 for more deets.

Comment (MetRonnie) [reply src/components/cylc/Info.vue:98]: `cylc show` does not have `c` in it, just numbers

Comment (oliver-sanders) [reply src/components/cylc/Info.vue:98]: Interesting. The `c` prefix is present in the GraphQL response:

```graphql
query {
  workflows {
    taskProxies {
      id
      prerequisites {
        satisfied
        expression
        conditions {
          taskId
          message
          reqState
          exprAlias
          satisfied
        }
      }
    }
  }
}
```

E.G:

```
{
  "data": {
    "workflows": [
      {
        "taskProxies": [
          {
            "id": "~me/show/run1//1/b",
            "prerequisites": []
          },
          {
            "id": "~me/show/run1//1/x",
            "prerequisites": [
              {
                "satisfied": false,
                "expression": "c0 & (c1 | c2) & (c3 | c4)",
                "conditions": [
                  {
                    "taskId": "1/a",
                    "message": "satisfied naturally",
                    "reqState": "succeeded",
                    "exprAlias": "c0",
                    "satisfied": true
                  },
```

I guess we could strip the first character? Presumably `cylc show` does that itself?

Comment (oliver-sanders) [reply src/components/cylc/Info.vue:98]: Yep, `cylc show` strips the `c`:

https://github.com/cylc/cylc-flow/blob/b8be2e3ef5b37288df5de4d62642b0d84e8eb02d/cylc/flow/scripts/show.py#L315

I'll update this to do the same.

Comment (oliver-sanders) [reply src/components/cylc/Info.vue]: Done.

Comment (MetRonnie) [reply src/components/cylc/Info.vue]: Here is where the `c` comes from (suddenly remembered seeing this on my `cylc-remove` work): https://github.com/cylc/cylc-flow/blob/63a6a377982eb91875a260e6b734435b3ae00740/cylc/flow/prerequisite.py#L239

Don't know why we can't just remove it?

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue:40]: Sorted.

The data-store node standin thinggy was missing the `type` field which the menu apparently uses.

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue]: `cylc show` uses `string.strip('c')` and I've used `string.replace(/c/g, '')` here, so we should be safe to remove it.

Comment (oliver-sanders) [reply src/components/cylc/Info.vue]: For #1893, I think we should remove the "alias" as a user-facing thing and find another way to display it (we're not bound by plain text here).

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:89]: This one is inconsistent with the others
```suggestion
        <v-expansion-panel-title color="blue-grey-lighten-1">
```

Comment (oliver-sanders) [reply src/components/cylc/Info.vue:89]: On purpose.

There will be other sections in the near future. I tried adding a few extras and it ended up as a giant block of "blue-grey" which didn't make it especially obvious that these are individual sections that can be expanded.

I thought it was clearer if the sections were striped ("lighten-1", "lighten-2", "lighten-1", ...).

File: src/services/mock/json/infoView.json
Comment (MetRonnie) [inline src/services/mock/json/infoView.json:48]: Typo
```suggestion
                  "exprAlias": "c0",
                  "satisfied": true
                },
                {
                  "taskId": "20000102T0000Z/eventually_succeeded",
                  "reqState": "succeeded",
                  "exprAlias": "c1",
```

File: src/services/mock/json/infoView.json
Comment (MetRonnie) [inline src/services/mock/json/infoView.json:29]: Getting a console warning
```
[Vue warn]: Invalid prop: type check failed for prop "startTime". Expected String with value "0", got Number with value 0. 
```

```suggestion
              "startedTime": "0",
```

File: src/components/cylc/Info.vue
Comment (MetRonnie) [reply src/components/cylc/Info.vue:89]: To me it doesn't really indicate clickability, it looks like it's selected or has keyboard focus or something. Not too important

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:159]: Any need to have this as a prop?

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:168]: (Funny, now that I am used to using `setup` composition API, having separate sections for `computed`, `methods` etc seems rather arbitrary)

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue:159]: This enables clean separation of the component and view.

The view handles data provision, configuration, session restoration, etc.

The component just renders the content.

**Comment by oliver-sanders** [comment]
Unfortunately, this work may be blocked by https://github.com/cylc/cylc-flow/issues/6307 which also imposes some back-compatibility constraints :(

I may be able to hack around it with special logic for handling the `prerequisites` and `outputs` fields as a workaround though.

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:214]: https://github.com/oliver-sanders/cylc-ui/pull/113

> Had a little play about - I like `mdiProgressHelper` as an icon to indicate outputs that haven't been completed yet
>
> image

**Comment by oliver-sanders** [comment]
We could overcome the block by moving from a delta subscription to a regular subscription.

However, this would request the complete data for the task on every change.

File: src/components/cylc/Info.vue
Comment (MetRonnie) [inline src/components/cylc/Info.vue:209]: To allow to text to wrap instead of overflowing
```suggestion
      white-space: pre-wrap;
```

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue]: Used filled and unfilled circles in native CSS/unicode (avoids need for surplus Vue components) to match the Cylc task icon convention.

**Comment by oliver-sanders** [comment]
The cylc-flow blocker to this has not been resolved, however, I have come up with a workaround: https://github.com/cylc/cylc-flow/pull/6403

I have also added an output completion expression so this should now be more or less a direct analogue of `cylc show` with a similar display interface.

File: src/components/cylc/Info.vue
Comment (oliver-sanders) [reply src/components/cylc/Info.vue]: And using one line of CSS avoids the need for a component?!

Comment (MetRonnie) [reply src/components/cylc/Info.vue]: It's 10 lines of CSS + 4 lines of template, vs 11 lines of template + 2 imports: https://github.com/oliver-sanders/cylc-ui/pull/113/commits/61f46c0c441be5589f2b59995af7f98210a7c05f#diff-11a1594bf34a47d021d9d6a0cfe88f5b235480cba1d7cb743a4ad7134238a594

Comment (oliver-sanders) [reply src/components/cylc/Info.vue]: eh? +25/-20!

**Comment by MetRonnie** [comment]
Have you/are you considering a client-side workaround (i.e. in the `onUpdated` method of the `InfoCallback`) rather than https://github.com/cylc/cylc-flow/pull/6403?

Other feedback: to me, these look like bullet points rather than yes/no indicators. (Tested on Windows, FF & Chrome)

image

**Comment by oliver-sanders** [comment]
> Have you/are you considering a client-side workaround 

It's possible, but it would require iterating over the received delta and the fix would only apply to the info view and not to other views which might want this information (e.g. the tree view currently uses it).

I think the long-term fix should be cylc-flow end so would rather stick any bodges / workarounds there if possible. Happy to undraft the hack PR, was hoping we could come up with a more global fix.

**Comment by hjoliver** [comment]
Unblocked by merge of https://github.com/cylc/cylc-flow/pull/6403 ... I think

File: cypress/component/info.cy.js
Comment (MetRonnie) [inline cypress/component/info.cy.js:180]: - These `expect(selector[0]).to.contain('N')` were all testing the topmost item in the list
- Less brittle to check for/against the existence of the particular class rather than the whole class list
```suggestion
        expect(selector[0].innerText).to.equal('(0 & 1) | 2')
        expect(selector[0]).to.not.have.class('satisfied')

        expect(selector[1].innerText).to.equal('0 a:succeeded')
        expect(selector[1]).to.have.class('satisfied')

        expect(selector[2].innerText).to.equal('1 b:custom_output')
        expect(selector[2]).to.not.have.class('satisfied')

        expect(selector[3].innerText).to.equal('2 a:expired')
        expect(selector[3]).to.not.have.class('satisfied')

        expect(selector[4].innerText).to.equal('0')
        expect(selector[4]).to.have.class('satisfied')

        expect(selector[5].innerText).to.equal('0 x:succeeded')
        expect(selector[5]).to.have.class('satisfied')
```

File: src/components/cylc/commandMenu/Menu.vue
Comment (MetRonnie) [inline src/components/cylc/commandMenu/Menu.vue:225]:
```suggestion
    isEditable (mutation = {}, authorised = false) {
      return !['log', 'info'].includes(mutation.name) && !this.isDisabled(mutation, authorised)
```

File: tests/e2e/specs/info.cy.js
Comment (MetRonnie) [inline tests/e2e/specs/info.cy.js:28]: Can make these less brittle
```suggestion
      // click on task
      .get('.node-data-task [data-c-interactive]:first')
      .click({ force: true })

      // from the menu select the "Info" psudo-mutation
      .get('.v-list-item')
```

File: src/views/views.js
Comment (MetRonnie) [inline src/views/views.js:77]: Note uses of `allViews` elsewhere. The info view shouldn't be selectable as the default view in settings: 
image

**Review by MetRonnie — APPROVED** [review]
Approved although ideally pending the fix for default view selection

**Comment by wxtim** [comment]
Re-reviewed and happy. 👍🏼

**Pull request merged**
